File: 3rd_party/adios/bindings/CXX11/adios2/cxx11/Attribute.cpp
/*
 * Distributed under the OSI-approved Apache License, Version 2.0.  See
 * accompanying file Copyright.txt for details.
 *
 * Attribute.cpp :
 *
 *  Created on: Jun 4, 2018
 *      Author: William F Godoy godoywf@ornl.gov
 */

#include "Attribute.h"

#include "adios2/common/ADIOSMacros.h"
#include "adios2/core/Attribute.h"
#include "adios2/helper/adiosFunctions.h"

namespace adios2
{

#define declare_type(T)                                                                            \
                                                                                                   \
    template <>                                                                                    \
    Attribute<T>::Attribute(core::Attribute<IOType> *attribute) : m_Attribute(attribute)           \
    {                                                                                              \
    }                                                                                              \
                                                                                                   \
    template <>                                                                                    \
    Attribute<T>::operator bool() const noexcept                                                   \
    {                                                                                              \
        return (m_Attribute == nullptr) ? false : true;                                            \
    }                                                                                              \
                                                                                                   \
    template <>                                                                                    \
    std::string Attribute<T>::Name() const                                                         \
    {                                                                                              \
        helper::CheckForNullptr(m_Attribute, "in call to Attribute<T>::Name()");                   \
        return m_Attribute->m_Name;                                                                \
    }                                                                                              \
                                                                                                   \
    template <>                                                                                    \
    std::string Attribute<T>::Type() const                                                         \
    {                                                                                              \
        helper::CheckForNullptr(m_Attribute, "in call to Attribute<T>::Type()");                   \
        return ToString(m_Attribute->m_Type);                                                      \
    }                                                                                              \
                                                                                                   \
    template <>                                                                                    \
    std::vector<T> Attribute<T>::Data() const                                                      \
    {                                                                                              \
        helper::CheckForNullptr(m_Attribute, "in call to Attribute<T>::Data()");                   \
                                                                                                   \
        if (m_Attribute->m_IsSingleValue)                                                          \
        {                                                                                          \
            return std::vector<T>{m_Attribute->m_DataSingleValue};                                 \
        }                                                                                          \
        else                                                                                       \
        {                                                                                          \
            return helper::NewVectorType<IOType, T>(m_Attribute->m_DataArray);                     \
        }                                                                                          \
    }                                                                                              \
                                                                                                   \
    template <>                                                                                    \
    bool Attribute<T>::IsValue() const                                                             \
    {                                                                                              \
        helper::CheckForNullptr(m_Attribute, "in call to Attribute<T>::IsValue()");                \
        return m_Attribute->m_IsSingleValue;                                                       \
    }                                                                                              \
                                                                                                   \
    template <>                                                                                    \
    std::string ToString(const Attribute<T> &attribute)                                            \
    {                                                                                              \
        return std::string("Attribute<") + attribute.Type() + ">(Name: \"" + attribute.Name() +    \
               "\")";                                                                              \
    }

ADIOS2_FOREACH_ATTRIBUTE_TYPE_1ARG(declare_type)
#undef declare_type

} // end namespace adios2

File: 3rd_party/adios/python/adios2/attribute.py
"""License:
  Distributed under the OSI-approved Apache License, Version 2.0.  See
  accompanying file Copyright.txt for details.
"""


class Attribute:
    """High level representation of the Attribute class in the adios2.bindings"""

    def __init__(self, io, name, *args, **kwargs):
        self.impl = io.DefineAttribute(name, *args, **kwargs)

    @property
    def impl(self):
        """Bindings implementation of the class"""
        return self._impl

    @impl.setter
    def impl(self, implementation):
        self._impl = implementation

    def __eq__(self, other):
        if isinstance(other, Attribute):
            return self.name() == other.name()
        return False

    def name(self):
        """
        Name of the Attribute

        Returns:
            Name of the Attribute as a str.
        """
        return self.impl.Name()

    def type(self):
        """
        Type of the Attribute

        Returns:
            Type of the Attribute as a str.
        """
        return self.impl.Type()

    def single_value(self):
        """
        True if the attribute is a single value, False if it is an array

        Returns:
            True or False.
        """
        return self.impl.SingleValue()

    def data(self):
        """
        Content of the Attribute

        Returns:
            Content of the Attribute as a non string.
        """
        if self.single_value():
            return self.impl.Data()[0]
        return self.impl.Data()

    def data_string(self):
        """
        Content of the Attribute

        Returns:
            Content of the Attribute as a str.
        """
        if self.single_value():
            return self.impl.DataString()[0]
        return self.impl.DataString()
